app/ota_client.py

```diff
@@ -314,6 +386,116 @@ def acquire_staging_storage(self):
             self._lock.release()
 
 
+class OtaStateSync:
+    """State machine that synchronzing ota_service and ota_client.
+
+    Typical usage:
+    a. wait on specific state
+        fsm: OtaStateSync
+        fsm.wait_on(fsm._S0, timeout=6)
+    b. expect <state>, and doing something to switch to next state
+        fsm: OtaStateSync
+        with fsm.proceed(fsm._P1, expect=fsm._START) as _next:
+            # do something here...
+            print(f"done! switch to next state {_next}")
+    """
+
+    ######## state machine definition ########
+    # states definition
+    _START, _S0, _S1, _S2, _END = (
+        "_START",  # start
+        "_S0",  # cache_scrub_finished
+        "_S1",  # pre_update_finished
+        "_S2",  # apply_update_finished
+        "_END",  # end
+    )
+    _STATE_LIST = ["_START", "_S0", "_S1", "_S2", "_END"]
```

Comment:
It is better not to use immediate value:
```suggestion
    _STATE_LIST = [_START, _S0, _S1, _S2, _END]
```